pkg/apiserver/rest/dualwriter.go

```diff
@@ -174,58 +167,69 @@ func SetDualWritingMode(
 	errDualWriterSetCurrentMode := errors.New("failed to set current dual writing mode")
 
 	// Use entity name as key
-	m, ok, err := kvs.Get(ctx, entity)
+	m, ok, err := kvs.Get(ctx, cfg.Kind)
 	if err != nil {
 		return Mode0, errors.New("failed to fetch current dual writing mode")
 	}
 
-	currentMode, valid := toMode[m]
+	currentMode, exists := toMode[m]
 
-	if !valid && ok {
+	// If the mode does not exist in our mapping, we log an error.
+	if !exists && ok {
 		// Only log if "ok" because initially all instances will have mode unset for playlists.
-		klog.Infof("invalid dual writing mode for %s mode: %v", entity, m)
+		klog.Infof("invalid dual writing mode for %s mode: %v", cfg.Kind, m)
 	}
 
-	if !valid || !ok {
+	// If the mode does not exist in our mapping, and we also didn't find an entry for this kind, fallback.
+	if !exists || !ok {
 		// Default to mode 1
 		currentMode = Mode1
-
-		err := kvs.Set(ctx, entity, fmt.Sprint(currentMode))
-		if err != nil {
+		if err := kvs.Set(ctx, cfg.Kind, fmt.Sprint(currentMode)); err != nil {
 			return Mode0, errDualWriterSetCurrentMode
 		}
 	}
 
+	// Handle transitions to the desired mode.
 	switch {
-	case desiredMode == Mode2 || desiredMode == Mode1:
-		currentMode = desiredMode
-		err := kvs.Set(ctx, entity, fmt.Sprint(currentMode))
-		if err != nil {
+	case cfg.Mode == Mode2 || cfg.Mode == Mode1:
+		// Directly set the mode for Mode1 and Mode2.
+		currentMode = cfg.Mode
+		if err := kvs.Set(ctx, cfg.Kind, fmt.Sprint(currentMode)); err != nil {
 			return Mode0, errDualWriterSetCurrentMode
 		}
-	case desiredMode >= Mode3 && currentMode < Mode3:
-		syncOk, err := runDataSyncer(ctx, currentMode, legacy, storage, entity, reg, serverLockService, requestInfo)
+	case cfg.Mode >= Mode3 && currentMode < Mode3:
+		// Transitioning to Mode3 or higher requires data synchronization.
+		cfgModeTmp := cfg.Mode
+		// Before running the sync, set the syncer config to the current mode, as we have to run the syncer
+		// once in the current active mode before we can upgrade.
+		cfg.Mode = currentMode
+		syncOk, err := runDataSyncer(ctx, cfg)
+		// Once we are done with running the syncer, we can change the mode back on the config to the desired one.
+		cfg.Mode = cfgModeTmp
 		if err != nil {
 			klog.Info("data syncer failed for mode:", m)
-			return currentMode, err
+			return Mode0, err
```

Comment:
Do we want to go back to Mode0 if the syncer fails? Or just default back to the current mode since it would be less than 3?

Reply:
If we return an error, the mode is not used that is returned by the function. I tried to make this more clear by defaulting to mode0 for all error cases as it's not nillable.